workflows/nsistp/shared/forms.py

```diff
@@ -0,0 +1,219 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import re
+from collections.abc import Iterator
+from datetime import datetime
+from functools import partial
+from typing import Annotated, Any
+from uuid import UUID
+
+from annotated_types import BaseMetadata, Ge, Le
+from orchestrator.db import ProductTable, db
+from orchestrator.db.models import (
+    SubscriptionTable,
+)
+from orchestrator.domain.base import SubscriptionModel
+from orchestrator.types import SubscriptionLifecycle
+from pydantic import AfterValidator, Field, ValidationInfo
+from pydantic_forms.types import UUIDstr
+from pydantic_forms.validators import Choice
+from sqlalchemy import select
+from typing_extensions import Doc
+
+from products.product_blocks.port import PortMode
+from products.product_types.nsistp import Nsistp, NsistpInactive
+from utils.exceptions import (
+    DuplicateValueError,
+    FieldValueError,
+)
+from workflows.nsistp.shared.shared import (
+    MAX_SPEED_POSSIBLE,
+    CustomVlanRanges,
+)
+from workflows.shared import subscriptions_by_product_type_and_instance_value
+
+TOPOLOGY_REGEX = r"^[-a-z0-9+,.;=_]+$"
+STP_ID_REGEX = r"^[-a-z0-9+,.;=_:]+$"
+NURN_REGEX = r"^urn:ogf:network:([^:]+):([0-9]+):([a-z0-9+,-.:;_!$()*@~&]*)$"
+FQDN_REQEX = (
+    r"^(?!.{255}|.{253}[^.])([a-z0-9](?:[-a-z-0-9]{0,61}[a-z0-9])?\.)*[a-z0-9](?:[-a-z0-9]{0,61}[a-z0-9])?[.]?$"
+)
+
+
+def port_selector() -> type[Choice]:
+    port_subscriptions = subscriptions_by_product_type_and_instance_value(
+        "Port", "port_mode", PortMode.TAGGED, [SubscriptionLifecycle.ACTIVE]
+    )
+    ports = {
+        str(subscription.subscription_id): subscription.description
+        for subscription in sorted(port_subscriptions, key=lambda port: port.description)
+    }
+    return Choice("Port", zip(ports.keys(), ports.items()))
+
+
+def is_fqdn(hostname: str) -> bool:
+    return re.match(FQDN_REQEX, hostname, re.IGNORECASE) is not None
+
+
+def valid_date(date: str) -> tuple[bool, str | None]:
+    def valid_month() -> tuple[bool, str | None]:
+        month_str = date[4:6]
+        month = int(month_str)
+        if month < 1 or month > 12:
+            return False, f"{month_str} is not a valid month number"
+        return True, None
+
+    def valid_day() -> tuple[bool, str | None]:
+        try:
+            datetime.fromisoformat(f"{date[0:4]}-{date[4:6]}-{date[6:8]}")
+        except ValueError:
+            return False, f"`{date}` is not a valid date"
+        return True, None
+
+    length = len(date)
+    if length == 4:  # year
+        pass  # No checks on reasonable year, so 9999 is allowed
+    elif length in (6, 8):
+        valid, message = valid_month()
+        if not valid:
+            return valid, message
+        if length == 8:  # year + month + day
+            return valid_day()
+    else:
+        return False, f"date `{date}` has invalid length"
+
+    return True, None
```

Comment:
This could rely entirely on the validation by `datetime.datetime.strptime`

```py
VALID_DATE_FORMATS = {
    4: '%Y', 
    6: '%Y%m', 
    8: '%Y%m%d',
}

def valid_date(date: str) -> tuple[bool, str | None]:
    if not (date_format := VALID_DATE_FORMATS.get(len(date))):
        return False, f"Invalid date length, expected one of: {list(VALID_DATE_FORMATS)}"

    try:
        _ = datetime.strptime(date, date_format)
        return True, None
    except ValueError as exc:
        return False, f"Invalid date for format {date_format!r}: {exc}"
```